Add graph view with custom revset

HKalbasi opened the pull request:
Ability of querying commits is a nice feature of jj, and not being a fan of terminal graphs, I decided to bring it to jjk. This PR adds another graph view which accepts custom revset (can accept more configuration in future) and appears in a bigger screen, and also add a `File History` menu item on top of that:

![image](https://github.com/user-attachments/assets/9f010fa9-ace1-4027-9a03-3b08085c03da)

kevin314 commented:
Wow! Thank you for your effort on this! I'll be sure to checkout these changes and test it out!

kevin314 commented:
I'm amenable to getting this merged, as it's very cool, but I wanted to let you know that we're thinking of revamping the graph view again. Something that would let us rebase easily, and maybe improve the clarity on bookmarks and branches. That might not be for a bit though--definitely until we fix the many bugs currently affecting JJK's basic functionality. That being said, I can see these features eventually being added to the new view as well. Thanks again for your work!

kevin314 commented:
I'll drop some review comments soon after I've tested some more!

HKalbasi commented:
>  we're thinking of revamping the graph view again. Something that would let us rebase easily, and maybe improve the clarity on bookmarks and branches

That would be great!

>  I can see these features eventually being added to the new view as well.

I tried to reuse code from the current graph view as much as possible, so ideally it will appear in the new view without too much effort.

HKalbasi commented:
@kevin314 what is the next step on this PR? Do you want to do the revamping before merging this?

kevin314 commented:
Sorry for the delay on this!

Here's some behavior I noticed:
1.  Our graph seems to not work well with filesets (i.e. when opening file history in the expanded graph) and a lot of revsets in general. It's most likely an issue with how we parse the cli output. Currently it doesn't really expect or handle a lot of formatting/syntax that's possible from jj log.
2. Periodic refreshing of the resource states/graph seems to be resetting the scrolling and clearing selections in the expanded graph.
3. Expanded graph doesn't save your provided revset if you switch tabs in the editor.
4. Sometimes the graph will alternate between different states when given a new revset.

The biggest issue is #1, but I think that's more due to a limitation of the current graph. We can try to address that first, but it'll involve a pretty big change..

https://github.com/user-attachments/assets/c8d462cc-9cc1-4649-ac4c-f3208d602d31

HKalbasi commented:
> Our graph seems to not work well with filesets (i.e. when opening file history in the expanded graph) and a lot of revsets in general. It's most likely an issue with how we parse the cli output. Currently it doesn't really expect or handle a lot of formatting/syntax that's possible from jj log.

How I can reproduce this? Does it happen on the jjk or jj repository?

I will try to fix other bugs you found. Thanks for testing.

HKalbasi commented:
I think we can use this template instead of parsing the `jj log` output manually:
```
jj log --no-graph --no-pager -T '"{ \"id\": " ++ stringify(change_id).escape_json() ++ ", \"description\": " ++ description.first_line().escape_json() ++ ", \"email\": " ++ stringify(author.email()).escape_json() ++ ", \"parents\": [" ++ parents.map(|c| stringify(c.change_id()).escape_json()).join(",") ++ "] }\n"'
```
Which returns a jsonl like this:
```
{
  "id": "rwzrwrkmzyvwukssyyspkukzpopnxopx",
  "description": "Add getStatus method to prevent double trigger of SCM refresh",
  "email": "kevlin3141@gmail.com",
  "parents": [
    "pxlxrwwkpxxpxynokwvzyqoryyputnzm"
  ]
}
{
  "id": "nsmpopxkxnlmwttqonpknznptsqukyqp",
  "description": "Fix showing absolute path in source control tree view",
  "email": "hamidrezakalbasi@protonmail.com",
  "parents": [
    "znktzwykyooyqrsowqqlpnlkmyukvqtr"
  ]
}
{
  "id": "znktzwykyooyqrsowqqlpnlkmyukvqtr",
  "description": "Specify revision when calling jj file annotate",
  "email": "keanemind@gmail.com",
  "parents": [
    "usptvyvolupnqpwoznlwxqrrurrnqomx"
  ]
}
```
But the problem is that it requires jj 0.27. Is there any requirement on supported versions of jj in jjk for now?

kevin314 commented:
I feel enforcing jj >= 0.27 is a little too strict at the moment, since its still the latest release. We should at the very least support the latest couple of versions of jj. Or, if we can't find another way to cleanly parse, maybe we only enable this feature for users with version 0.27

HKalbasi commented:
If the bar is at supporting latest couple of versions (which I think is a reasonable bar, specially at this stage of jj and jjk development) we can just wait two weeks for 0.28 to be released. We can think about other ways to parse log or support conditioning on jj version, but I don't think it worth complicating the code.

Until jj 0.28 is released, I will try to fix other bugs you found.

lishaduck commented:
> we can just wait two weeks for 0.28 to be released

It's out :)

HKalbasi commented:
@kevin314 I fixed the problem 1 by switching the parser to the json based template, and it seems it is now working. About problem 2, I was wrongly used `force = true` every time. About the problem 3, I didn't do any specific thing for it, and I was able to reproduce it previously, but after rebase and updating vscode, it seems is now fixed. I was not able to reproduce problem 4.

kevin314 commented:
Thanks for the change, and apologies for the delay! When I pulled your latest changes to test, I noticed some inconsistent behavior with how my graph is rendered--this is what the jj repo looks like on your branch. On the right side in the terminal is the output of `jj log -r ::` 

![image](https://github.com/user-attachments/assets/dbb83745-520e-4724-b177-c40e7fe2f2af)

For reference, this is what it looks like on main:

![image](https://github.com/user-attachments/assets/0a97d35d-878b-43f7-a3fa-c0d8105c836e)


Are you able to recreate this behavior at all?

kevin314 commented:
Thanks! I took another look and noticed a couple more things. 

Switching to a change no longer seems to visually update the expanded revision graph properly anymore, i.e, using either the regular graph or the expanded graph will move to the correct change in the backend, but only the regular graph's UI will update to reflect the change.

Also, I noticed the graph is now using ◆ instead of • as delimiters, was this intended?

HKalbasi commented:
> Also, I noticed the graph is now using ◆ instead of • as delimiters, was this intended?

No it was not, I just copied it from wrong place. I fixed it.

> Switching to a change no longer seems to visually update the expanded revision graph properly anymore, i.e, using either the regular graph or the expanded graph will move to the correct change in the backend, but only the regular graph's UI will update to reflect the change.

This is sad. The problem is that, `updateResources` is called every ~5 seconds, even if there is no events, and this is the one responsible for updating the graphs. In that function, if I use refresh with `force=true`, there will be loss of scroll and selections, and if I use `force = false`, then it could miss some changes. What should I do?
